### ISSUE TITLE
Focus and Redraw Systems

### ISSUE THREAD

**Issue by rushsteve1** [redox-os/orbtk#28]
The current system of determining focus and determining if an widget needs a redraw is really problematic. It is convoluted and almost unreadable, as well as just being poor code:
```rust
let mut redraw = false;
for event in events.iter() {
    for i in 0..self.widgets.borrow().len() {
        if let Some(widget) = self.widgets.borrow().get(i) {
            if widget.event(*event, self.widget_focus.get() == i, &mut redraw) {
                if self.widget_focus.get() != i {
                    self.widget_focus.set(i);
                    redraw = true;
                }
            }
        }
    }
}

if redraw {
    self.draw();
}
```
Iterating through the list of widgets with the `widget_focus` field on the `Window` struct as the index of the focused widget is bad form. And using the `redraw` variable in that way is strange to say the least.

The issues with this system is primarily that a widget cannot be redrawn without stealing focus. the `self.widget_focus.set(i)` immediately sets the focus to be the widget that needs a redraw. Additionally the `draw()` function on the `Window` struct has semi-redundant code:
```rust
for i in 0..self.widgets.borrow().len() {
    if let Some(widget) = self.widgets.borrow().get(i) {
        widget.draw(&mut renderer, self.widget_focus.get() == i);
    }
}
```

The focus and redraw systems need to be separated out. A containers based focus and redraw systems would also go well with #10.

**Comment by jackpot51** [comment]
> The issues with this system is primarily that a widget cannot be redrawn without stealing focus.
Not true. The widget can set redraw to true and return false. This code just means that upon stealing focus, the widget is forced to redraw.

Another need is for a widget to focus automatically, such as the username in `orblogin`. I agree that we should refactor the way focus works, I just don't know what that should look like.

**Comment by NilSet** [comment]
I was thinking of having a `fn focus(Arc<Widget>)` function to send focus to another element.

**Comment by rushsteve1** [comment]
I like @NilSet's idea. Additionally I think that `widget_focus` should be some kind of reference to the widget itself.

**Comment by FloVanGH** [comment]
Implemented a base version of a FocusManager with #74. 

https://github.com/redox-os/orbtk/blob/master/src/focus_manager.rs
